hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/HapiSuite.java

```diff
@@ -285,13 +297,78 @@ private static void runSequentialSpecs(final List<HapiSpec> specs) {
+                .map(r -> CompletableFuture.runAsync(r, HapiSpec.getCommonThreadPool())
+                        .orTimeout(MAX_SPEC_RUNTIME_MINUTES, TimeUnit.MINUTES)
+                        .handleAsync(new HandleSpecCompletion(r), HapiSpec.getCommonThreadPool()))
+                .toArray(CompletableFuture[]::new);
+        CompletableFuture.allOf(futures)
+                .orTimeout(MAX_SUITE_RUNTIME_MINUTES, TimeUnit.MINUTES)
+                .join();
     }
 
     public static String salted(String str) {
         return str + RANDOM.nextInt(1_234_567);
     }
+
+    /**
+     * Basic BiFunction handler to ensure that spec failures are reported as soon as possible.
+     * <br/>This assumes the previous stage is an "<pre>orTimeout</pre>" stage so that when a spec
+     * takes too long to execute, it is stopped and reported rather than running forever.
+     * <p/>
+     * Currently, this just reports failures, as non-failure is adequately summarized elsewhere.
+     * Future work should remove the summary, and add result reporting, with full detail, here so that
+     * results are reported immediately for each spec in the suite rather than waiting until
+     * the entire suite is completed.
+     */
+    @SuppressWarnings("java:S106")
+    private static class HandleSpecCompletion implements BiFunction<Void, Throwable, Void> {
+        private static final String FULL_LOG_MESSAGE =
+                """
+                Spec %2$s failed exceptionally due to %3$s
+                Spec:
+                %1$s
+                Stack Trace:
+                %4$s
+                """;
+        private final HapiSpec specToHandle;
+
+        private HandleSpecCompletion(final HapiSpec specToHandle) {
+            this.specToHandle =
+                    Objects.requireNonNull(specToHandle, "Suite Attempted to run null HAPI spec concurrently.");
+        }
+
+        @Override
+        public Void apply(final Void result, final Throwable error) {
+            if (error != null) {
+                final Logger resultLogger = specToHandle.getResultLogger() == null
+                        ? LogManager.getLogger()
+                        : specToHandle.getResultLogger();
+                // write to stderr, because result log seems to get lost.
+                // Also write to the result log, for after that is fixed.
+                final String stackTrace = writeStackTraceToString(error);
+                final String detailedError =
+                        FULL_LOG_MESSAGE.formatted(specToHandle, specToHandle.getName(), error, stackTrace);
+                System.err.print(detailedError);
+                resultLogger.warn(detailedError);
+                throw new AssertionError(detailedError, error);
+            }
+            return result;
+        }
+
+        private static String writeStackTraceToString(final Throwable error) {
+            String stackTrace;
+            try (final StringWriter outputWriter = new StringWriter();
```

Comment:
(I wonder why the Java framework doesn't have a StringWriter constructor taking a StringBuilder, or something else useful like StringBuilder implementing interface Writer?)

Reply:
Only option is an int initial size (which isn't super helpful here).
As to why it was done this way, I couldn't say.  It is where it is.  It would be even easier if we could pass a StringBuilder to the Throwable printStackTrace, but we cannot do that either.